File: gitdot-web/packages/api/src/endpoint/repository/create-repository.ts
import { z } from "zod";
import { RepositoryResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const CreateRepositoryRequest = z.object({
  owner_type: z.string(),
  visibility: z.string(),
  description: z.string().optional(),
});
export type CreateRepositoryRequest = z.infer<typeof CreateRepositoryRequest>;

export const CreateRepositoryResponse = RepositoryResource;
export type CreateRepositoryResponse = z.infer<typeof CreateRepositoryResponse>;

export const CreateRepository = {
  path: "/repository/{owner}/{repo}",
  method: "POST",
  request: CreateRepositoryRequest,
  response: CreateRepositoryResponse,
} as const satisfies Endpoint;
export type CreateRepository = typeof CreateRepository;

File: gitdot-api/src/endpoint/repository/create_repository.rs
use serde::{Deserialize, Serialize};

use super::default_visibility;
use crate::{endpoint::Endpoint, resource::RepositoryResource};

pub struct CreateRepository;

impl Endpoint for CreateRepository {
    const PATH: &'static str = "/repository/{owner}/{repo}";
    const METHOD: http::Method = http::Method::POST;

    type Request = CreateRepositoryRequest;
    type Response = CreateRepositoryResponse;
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct CreateRepositoryRequest {
    pub owner_type: String,

    #[serde(default = "default_visibility")]
    pub visibility: String,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    #[serde(default)]
    pub init_readme: bool,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub gitignore_template: Option<String>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub license_template: Option<String>,
}

pub type CreateRepositoryResponse = RepositoryResource;
